File: litho-testing/src/main/java/com/facebook/litho/testing/LithoViewRule.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.testing;

import static android.view.View.MeasureSpec.EXACTLY;
import static android.view.View.MeasureSpec.UNSPECIFIED;
import static android.view.View.MeasureSpec.makeMeasureSpec;

import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.Nullable;
import androidx.test.core.app.ApplicationProvider;
import com.facebook.litho.Component;
import com.facebook.litho.ComponentContext;
import com.facebook.litho.ComponentTree;
import com.facebook.litho.ComponentsPools;
import com.facebook.litho.LayoutState;
import com.facebook.litho.LithoLayoutResult;
import com.facebook.litho.LithoView;
import com.facebook.litho.TreeProps;
import com.facebook.litho.annotations.TreeProp;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

/**
 * This test utility allows clients to test assertion on the view hierarchy rendered by a Litho
 * components. The utility has methods to override the default {@link LithoView}, {@link
 * ComponentTree}, width, and height specs.
 *
 * <pre><code>{@literal @RunWith(LithoTestRunner.class)}
 * public class LithoSampleTest {
 *
 *   public final @Rule LithoViewRule mLithoViewRule = new LithoViewRule();
 *
 *  {@literal @Test}
 *   public void test() {
 *     final ComponentContext c = mLithoViewRule.getContext();
 *     final Component component = MyComponent.create(c).build();
 *
 *     mLithoViewRule.setRoot(component)
 *       .attachToWindow()
 *       .setRoot(component)
 *       .measure()
 *       .layout();
 *
 *     LithoView lithoView = mLithoViewRule.getLithoView();
 *
 *     // Test your assertions on the litho view.
 *   }
 * }
 *
 * }</code></pre>
 */
public class LithoViewRule implements TestRule {

  public static final int DEFAULT_WIDTH_SPEC = makeMeasureSpec(1080, EXACTLY);
  public static final int DEFAULT_HEIGHT_SPEC = makeMeasureSpec(0, UNSPECIFIED);

  private ComponentContext mContext;
  private ComponentTree mComponentTree;
  private LithoView mLithoView;
  private int mWidthSpec = DEFAULT_WIDTH_SPEC;
  private int mHeightSpec = DEFAULT_HEIGHT_SPEC;

  @Override
  public Statement apply(final Statement base, Description description) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        try {
          mContext = new ComponentContext(ApplicationProvider.getApplicationContext());
          mContext.setLayoutStateContextForTesting();
          base.evaluate();
        } finally {
          ComponentsPools.clearMountContentPools();
          mContext = null;
          mComponentTree = null;
          mLithoView = null;
          mWidthSpec = DEFAULT_WIDTH_SPEC;
          mHeightSpec = DEFAULT_HEIGHT_SPEC;
        }
      }
    };
  }

  /** Gets the current {@link ComponentContext}. */
  public ComponentContext getContext() {
    return mContext;
  }

  public LithoViewRule useContext(ComponentContext c) {
    mContext = c;
    return this;
  }

  /** Gets the current {@link LithoView}; creates a new instance if {@code null}. */
  public LithoView getLithoView() {
    if (mLithoView == null) {
      mLithoView = new LithoView(mContext);
    }

    if (mLithoView.getComponentTree() == null) {
      mLithoView.setComponentTree(getComponentTree());
    }

    return mLithoView;
  }

  /** Sets a new {@link LithoView} which should be used to render. */
  public LithoViewRule useLithoView(LithoView lithoView) {
    mLithoView = lithoView;
    if (mLithoView.getComponentContext() != mContext) {
      throw new RuntimeException(
          "You must use the same ComponentContext for the LithoView as what is on the LithoViewRule @Rule!");
    }
    if (mComponentTree != null) {
      mLithoView.setComponentTree(mComponentTree);
    }
    return this;
  }

  /** Gets the current {@link ComponentTree}; creates a new instance if {@code null}. */
  public ComponentTree getComponentTree() {
    if (mComponentTree == null) {
      mComponentTree = ComponentTree.create(mContext).build();
    }
    return mComponentTree;
  }

  /** Sets a new {@link ComponentTree} which should be used to render. */
  public LithoViewRule useComponentTree(ComponentTree componentTree) {
    mComponentTree = componentTree;
    getLithoView().setComponentTree(componentTree);
    return this;
  }

  /** Sets the new root {@link Component} to render. */
  public LithoViewRule setRoot(Component component) {
    getComponentTree().setRoot(component);
    return this;
  }

  /** Sets the new root {@link Component.Builder} to render. */
  public LithoViewRule setRoot(Component.Builder builder) {
    getComponentTree().setRoot(builder.build());
    return this;
  }

  /** Sets the new root {@link Component} to render asynchronously. */
  public LithoViewRule setRootAsync(Component component) {
    getComponentTree().setRootAsync(component);
    return this;
  }

  /** Sets the new root {@link Component.Builder} to render asynchronously. */
  public LithoViewRule setRootAsync(Component.Builder builder) {
    getComponentTree().setRootAsync(builder.build());
    return this;
  }

  /** Sets the new root {@link Component} with new size spec to render. */
  public LithoViewRule setRootAndSizeSpec(Component component, int widthSpec, int heightSpec) {
    mWidthSpec = widthSpec;
    mHeightSpec = heightSpec;
    getComponentTree().setRootAndSizeSpec(component, mWidthSpec, mHeightSpec);
    return this;
  }

  /** Sets a new width and height which should be used to render. */
  public LithoViewRule setSizePx(int widthPx, int heightPx) {
    mWidthSpec = makeMeasureSpec(widthPx, EXACTLY);
    mHeightSpec = makeMeasureSpec(heightPx, EXACTLY);
    return this;
  }

  /** Sets a new width spec and height spec which should be used to render. */
  public LithoViewRule setSizeSpecs(int widthSpec, int heightSpec) {
    mWidthSpec = widthSpec;
    mHeightSpec = heightSpec;
    return this;
  }

  /** Sets a new {@link TreeProp} for the next layout pass. */
  public LithoViewRule setTreeProp(Class<?> klass, Object instance) {
    TreeProps props = mContext.getTreeProps();
    if (props == null) {
      props = new TreeProps();
    }

    props.put(klass, instance);

    mContext.setTreeProps(props);

    return this;
  }

  /** Explicitly calls measure on the current root {@link LithoView} */
  public LithoViewRule measure() {
    getLithoView().measure(mWidthSpec, mHeightSpec);
    return this;
  }

  /** Explicitly calls layout on the current root {@link LithoView} */
  public LithoViewRule layout() {
    final LithoView lithoView = getLithoView();
    lithoView.layout(0, 0, lithoView.getMeasuredWidth(), lithoView.getMeasuredHeight());
    return this;
  }

  /** Explicitly attaches current root {@link LithoView} */
  public LithoViewRule attachToWindow() {
    getLithoView().onAttachedToWindowForTest();
    return this;
  }

  /** Explicitly detaches current root {@link LithoView} */
  public LithoViewRule detachFromWindow() {
    getLithoView().onDetachedFromWindowForTest();
    return this;
  }

  /** Explicitly releases current root {@link LithoView} */
  public LithoViewRule release() {
    getLithoView().release();
    return this;
  }

  /** Gets the current width spec */
  public int getWidthSpec() {
    return mWidthSpec;
  }

  /** Gets the current height spec */
  public int getHeightSpec() {
    return mHeightSpec;
  }

  /**
   * Finds the first {@link View} with the specified tag in the rendered hierarchy, returning null
   * if is doesn't exist.
   */
  public @Nullable View findViewWithTagOrNull(Object tag) {
    return findViewWithTagTransversal(mLithoView, tag);
  }

  /**
   * Finds the first {@link View} with the specified tag in the rendered hierarchy, throwing if it
   * doesn't exist.
   */
  public View findViewWithTag(Object tag) {
    final View view = findViewWithTagOrNull(tag);
    if (view == null) {
      throw new RuntimeException("Did not find view with tag '" + tag + "'");
    }
    return view;
  }

  private @Nullable View findViewWithTagTransversal(View view, Object tag) {
    if (view.getTag() != null && view.getTag().equals(tag)) {
      return view;
    }
    if (view instanceof ViewGroup) {
      ViewGroup viewGroup = ((ViewGroup) view);
      for (int i = 0; i < viewGroup.getChildCount(); i++) {
        View child = findViewWithTagTransversal(viewGroup.getChildAt(i), tag);
        if (child != null) {
          return child;
        }
      }
    }
    return null;
  }

  protected @Nullable LayoutState getCommittedLayoutState() {
    return getComponentTree().getCommittedLayoutState();
  }

  public @Nullable LithoLayoutResult getCurrentRootNode() {
    return getCommittedLayoutState() != null ? getCommittedLayoutState().getLayoutRoot() : null;
  }
}

File: litho-testing/src/main/java/com/facebook/litho/testing/LithoViewRule.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho.testing

import android.app.Activity
import android.content.Context
import android.os.Looper
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import com.facebook.litho.Component
import com.facebook.litho.ComponentContext
import com.facebook.litho.ComponentScope
import com.facebook.litho.ComponentTree
import com.facebook.litho.LithoLifecycleProvider
import com.facebook.litho.LithoView
import com.facebook.litho.TreeProps
import com.facebook.litho.config.ComponentsConfiguration
import com.facebook.rendercore.MountItemsPool
import com.facebook.rendercore.utils.MeasureSpecUtils.exactly
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement
import org.robolectric.Robolectric
import org.robolectric.Shadows.shadowOf

/**
 * This test utility allows clients to create a [TestLithoView] instance that allows to test
 * assertion on the view hierarchy rendered by a Litho components.
 *
 * ```
 *  @RunWith(AndroidJUnit4::class)
 *  class LithoSampleTest {
 *
 *  @Rule @JvmField val lithoViewRule = LithoViewRule()
 *  @Test
 *  fun test() {
 *   val testLithoView =  lithoViewRule.render { TestComponent() }
 *
 *    // or you can use setRoot/measure/layout for more fine-grained control
 *    val testLithoView = lithoViewRule.createTestLithoView().attachToWindow().setRoot(TestComponent()).measure().layout()
 *    // Test your assertions on the TestLithoView instance.
 *    }
 * }
 * ```
 */
class LithoViewRule
@JvmOverloads
constructor(
    val componentsConfiguration: ComponentsConfiguration? = null,
    val themeResId: Int? = null,
    private val lithoLifecycleProvider: (() -> LithoLifecycleProvider)? = null
) : TestRule {
  lateinit var context: ComponentContext
  private var threadLooperController: BaseThreadLooperController = ThreadLooperController()

  init {
    ComponentsConfiguration.isDebugModeEnabled = true
  }

  override fun apply(base: Statement, description: Description): Statement {
    return object : Statement() {
      override fun evaluate() {
        ensureThreadLooperType()

        try {
          if (themeResId != null) {
            val activity = Robolectric.buildActivity(Activity::class.java).create().get()
            activity.setTheme(themeResId)
            context = ComponentContext(activity)
          } else {
            context = ComponentContext(getApplicationContext<Context>())
          }

          threadLooperController.init()
          base.evaluate()
        } finally {
          threadLooperController.clean()
          MountItemsPool.clear()
          context.clearCalculationStateContext()
        }
      }
    }
  }

  private fun ensureThreadLooperType() {
    if (ComponentsConfiguration.isSplitResolveAndLayoutWithSplitHandlers() &&
        threadLooperController is ThreadLooperController) {
      threadLooperController = ResolveAndLayoutThreadLooperController()
    } else if (!ComponentsConfiguration.isSplitResolveAndLayoutWithSplitHandlers() &&
        threadLooperController is ResolveAndLayoutThreadLooperController) {
      threadLooperController = ThreadLooperController()
    }
  }

  fun useContext(c: ComponentContext): LithoViewRule {
    context = c
    return this
  }

  /** Sets a new [TreeProp] for the next layout pass. */
  fun setTreeProp(klass: Class<*>, instance: Any?): LithoViewRule {
    val props = context.treeProps ?: TreeProps()
    props.put(klass, instance)
    context.treeProps = props
    return this
  }

  /**
   * Creates new TestLithoView holder responsible for keeping an instance of LithoView, allowing to
   * find Views/Components or perform assertions on it. For simple Component rendering without
   * fine-grained control, use [render]
   */
  @JvmOverloads
  fun createTestLithoView(
      lithoView: LithoView? = null,
      componentTree: ComponentTree? = null,
      widthPx: Int? = null,
      heightPx: Int? = null,
      componentFunction: (ComponentScope.() -> Component?)? = null
  ): TestLithoView {
    val testLithoView =
        TestLithoView(context, componentsConfiguration, lithoLifecycleProvider?.invoke())
    componentTree?.let { testLithoView.useComponentTree(componentTree) }
    lithoView?.let { testLithoView.useLithoView(lithoView) }
    if (widthPx != null || heightPx != null) {
      val widthSpec = if (widthPx != null) exactly(widthPx) else DEFAULT_WIDTH_SPEC
      val heightSpec = if (heightPx != null) exactly(heightPx) else DEFAULT_HEIGHT_SPEC
      testLithoView.setSizeSpecs(widthSpec, heightSpec)
    }

    componentFunction?.let {
      with(ComponentScope(context)) { componentFunction() }
          ?.let { component -> testLithoView.setRoot(component) }
    }
    return testLithoView
  }

  /** Sets the new root to render. */
  @JvmOverloads
  fun render(
      lithoView: LithoView? = null,
      componentTree: ComponentTree? = null,
      widthPx: Int? = null,
      heightPx: Int? = null,
      componentFunction: ComponentScope.() -> Component?
  ): TestLithoView {
    val testLithoView =
        createTestLithoView(
            lithoView = lithoView,
            componentTree = componentTree,
            widthPx = widthPx,
            heightPx = heightPx,
            componentFunction = componentFunction)
    return testLithoView.attachToWindow().measure().layout()
  }

  /**
   * Perform any interactions defined in the [InteractionScope] or on the [LithoViewRule].
   *
   * During tests we need to make sure that everything is in sync in the Main Thread and in the
   * Background Thread, just like in real life use case. This functions takes off the responsibility
   * from you to use the Loopers and manage the thread synchronisation. You only need to pass here
   * one of the defined interactions from [LithoViewRule] or [InteractionScope], and we will take
   * care of all of the rest
   */
  fun act(
      testLithoView: TestLithoView,
      action: TestLithoView.InteractionsScope.() -> Unit
  ): LithoViewRule {
    testLithoView.InteractionsScope().action()
    idle()
    return this
  }

  /**
   * Runs through all tasks on the background thread and main lopper, blocking until it completes.
   * Use if there are any async events triggered by layout ( ie visibility events) to manually drain
   * the queue
   */
  fun idle() {
    threadLooperController.runToEndOfTasksSync()
    shadowOf(Looper.getMainLooper()).idle()
  }
}
